### PR TITLE
LOOP-1869 Allow CGMManager to report valid session

### PR DESCRIPTION
**Pull request by ps2**
https://tidepool.atlassian.net/browse/LOOP-1869

### CODE REVIEW

File: LoopKit/DeviceManager/CGMManager.swift [hunk 1]
```diff
@@ -49,6 +57,12 @@ public protocol CGMManagerDelegate: DeviceManagerDelegate {
     /// - Parameter manager: The manager instance
     /// - Returns: The unique prefix for the credential store
     func credentialStoragePrefix(for manager: CGMManager) -> String
+    
+    /// Notifies the delegate of a change in status
+    ///
+    /// - Parameter manager: The manager instance
+    /// - Parameter status: The new, updated status
+    func cgmManager(_ manager: CGMManager, didUpdate status: CGMManagerStatus)
```
Comment (rickpasetto) [inline LoopKit/DeviceManager/CGMManager.swift:66]: I have to admit, this delegate is starting to get confusing.  There are three "did update" functions that seem very similar:

```swift
    func cgmManager(_ manager: CGMManager, didUpdateWith result: CGMResult) -> Void
```
```swift
    func cgmManagerDidUpdateState(_ manager: CGMManager)
```
...and now:
```swift
    func cgmManager(_ manager: CGMManager, didUpdate status: CGMManagerStatus)
```

Is there any way we could either coalesce them or rename them so they aren't so similar?

File: LoopKit/DeviceManager/CGMManager.swift [hunk 2]
```diff
@@ -19,6 +19,14 @@ public enum CGMResult {
     case error(Error)
 }
 
+public struct CGMManagerStatus {
```
Comment (rickpasetto) [inline LoopKit/DeviceManager/CGMManager.swift:22]: Why not add another `enum` to `CGMResult`, `hasValidSensorSession` with a `Bool` payload?  Wouldn't that obviate the need for an additional struct and `delegate` call?

File: LoopKit/GlucoseKit/GlucoseStore.swift [hunk 3]
```diff
@@ -286,6 +286,7 @@ extension GlucoseStore {
         dataAccessQueue.async {
             self.purgeCachedGlucoseObjects(matching: cachePredicate)
             self.healthStore.deleteObjects(of: self.glucoseType, predicate: healthKitPredicate, withCompletion: completion)
+            self.updateLatestGlucose()
```
Comment (rickpasetto) [inline LoopKit/GlucoseKit/GlucoseStore.swift]: Hm...how is this related to the valid sensor session?

File: LoopKit/GlucoseKit/GlucoseStore.swift [hunk 4]
```diff
@@ -474,6 +475,8 @@ extension GlucoseStore {
 
                 if let lastObject = objects.first {
                     self.latestGlucose = StoredGlucoseSample(managedObject: lastObject)
+                } else {
+                    self.latestGlucose = nil
```
Comment (rickpasetto) [inline LoopKit/GlucoseKit/GlucoseStore.swift]: Unit test for this?

File: LoopKit/InsulinKit/DoseStore.swift [hunk 5]
```diff
@@ -1165,7 +1165,8 @@ extension DoseStore {
                 // Return the larger of the two bounding values, for the scenario when a bolus
                 // was scheduled between the two values; we want to return the later, larger value
                 guard let maxValue = closest.max(by: { return $0.value < $1.value }) else {
-                    completion(.failure(.fetchError(description: "No values found", recoverySuggestion: "Ensure insulin data exists for the specified date")))
+                    // If we have no iob values in the store, and did not encounter an error, return 0
+                    completion(.success(InsulinValue(startDate: date, value: 0)))
```
Comment (rickpasetto) [inline LoopKit/InsulinKit/DoseStore.swift:1169]: Unit test for this?

**Review by rickpasetto — COMMENTED** [review]
Some questions for you before proceeding...

File: LoopKit/DeviceManager/CGMManager.swift (hunk 1 above)
Comment (ps2) [reply LoopKit/DeviceManager/CGMManager.swift:66]: Perhaps the most important part of reading these method signatures is the noun. And each one is different, and should indicate the scope of the update. If you have suggestions to improve the clarity of those nouns, I'm open, but I think the nouns are there and are what should distinguish them.
 
* CGMResult = a reading from the sensor
* CGMManagerStatus = status of the manager; includes properties that are not associated with a sensor reading. Battery level might go here, for example, if we need to report that, or transmitter/sensor lifetime estimations.
* State = the opaque state that Loop is responsible for persisting, and using to restore the CGMManager with.

The verb (didUpdate) can stay the same, to indicate that this call is triggered by some kind of change, I think.

I'll update the inline documentation to describe some of these things as well.

File: LoopKit/DeviceManager/CGMManager.swift (hunk 2 above)
Comment (ps2) [reply LoopKit/DeviceManager/CGMManager.swift:22]: I think it makes sense to keep updates that are not associated with individual sensor readings separate. The CGM result is used for reporting readings that are automatically read, and also when Loop wants to try to fetch readings. I think we should keep that field scoped to those actions.

File: LoopKit/GlucoseKit/GlucoseStore.swift (hunk 3 above)
Comment (ps2) [reply LoopKit/GlucoseKit/GlucoseStore.swift]: This is fixing a separate bug, that I noticed when working on this, where latestGlucose is not cleared out after testing sensor data is purged.

Comment (ps2) [reply LoopKit/GlucoseKit/GlucoseStore.swift]: I can put this into a separate PR if you feel that's warranted.

**Review by darinkrauss — APPROVED** [review]
LGTM! 🪀

File: LoopKit/DeviceManager/CGMManager.swift (hunk 1 above)
Comment (rickpasetto) [reply LoopKit/DeviceManager/CGMManager.swift:66]: Ok.  You have to admit the nouns are also not very distinguishing ("Result", "Status", "State").  Maybe consider:
`CGMResult` -> `CGMSensorReading` 
`CGMManagerStatus` ->  `CGMStatus` (it really is the status of the CGM, not of the Manager, right?)
I'll leave it up to you to rename, if you want.  Adding clearer documentation to the `protocol` functions is a good idea, please do.

File: LoopKit/DeviceManager/CGMManager.swift (hunk 2 above)
Comment (rickpasetto) [reply LoopKit/DeviceManager/CGMManager.swift:22]: I think this boils down to the above issue with `Result`, `Status`, and `State` being so similar.  Renaming them (or, at least, documenting [comments] exactly what their difference is) would help, I think.  Thanks.

File: LoopKit/GlucoseKit/GlucoseStore.swift (hunk 3 above)
Comment (rickpasetto) [reply LoopKit/GlucoseKit/GlucoseStore.swift]: Ah, nah, not necessary to split out into another PR. :)

FWIW, I'd be fine, next time, if you just added a PR comment before reviewing starts that says, "Yeah, this is a separate [unfiled] bug fix, I just rolled it in here for convenience."

**Review by rickpasetto — APPROVED** [review]
LGTM! (and thanks for explaining the different update calls)

File: LoopKit/DeviceManager/CGMManager.swift (hunk 1 above)
Comment (ps2) [reply LoopKit/DeviceManager/CGMManager.swift:66]: Yeah, I strongly considered a `CGMResult` -> `CGMSensorReading`, but it's not a reading because it also handles errors and missing readings, and is used for responses to queries to get readings, so it's more like a result that might or might not contain a reading.  Maybe CGMSensorReadingResult? I think that fits what it is. It's a little wordy, but I'm ok with that.

Comment (ps2) [reply LoopKit/DeviceManager/CGMManager.swift:66]: As for CGMManagerStatus -> CGMStatus, I probably would include any status changes related to the manager itself in this type, too, outside of the CGM hardware. But I'm fine with either name, if you still think CGMStatus is better.

Comment (rickpasetto) [reply LoopKit/DeviceManager/CGMManager.swift:66]: Hm...your comment about `CGMResult` seems to be an excellent case for `Result<>`:
Maybe:
```swift
// Before
public enum CGMResult {
    case noData
    case newData([NewGlucoseSample])
    case error(Error)
}
```
```swift
// After
public enum CGMError: Error {
    case noData
    case unknown(Error)
}

public typealias CGMSensorReadingResult = Result<[NewGlucoseSample], CGMError>
```

Just a suggestion...WDYT?

I don't feel as strongly about the change to `CGMManagerStatus`.  I have a slight preference for `CGMStatus`, but that's about it.

Comment (ps2) [reply LoopKit/DeviceManager/CGMManager.swift:66]: I think that just makes for a more nested result that is harder to work with. And .noData is not really a .failure. How about renaming CGMResult to CGMReadingResult? And the method could be 
`func cgmManager(_ manager: CGMManager, hasNew readingResult: CGMReadingResult) -> Void`

Comment (rickpasetto) [reply LoopKit/DeviceManager/CGMManager.swift:66]: Sure, that's ok with me.

**Review by rickpasetto — APPROVED** [review]
LGTM ⛓️

File: LoopKit/GlucoseKit/GlucoseStore.swift [hunk 6]
```diff
@@ -460,6 +465,35 @@ extension GlucoseStore {
...
+            self.cacheStore.managedObjectContext.performAndWait {
+                let request: NSFetchRequest<CachedGlucoseObject> = CachedGlucoseObject.fetchRequest()
+                request.predicate = NSPredicate(format: "startDate >= %@ AND wasUserEntered == NO", since as NSDate)
+                request.sortDescriptors = [NSSortDescriptor(key: "startDate", ascending: false)]
+                request.fetchLimit = 1
+
+                do {
+                    let objects = try self.cacheStore.managedObjectContext.fetch(request)
+                    let samples = objects.map { StoredGlucoseSample(managedObject: $0) }
+                    if samples.count > 0 {
```
Comment (darinkrauss) [inline LoopKit/GlucoseKit/GlucoseStore.swift]: [nit] Consider `completion(.success(samples.first))`

**Review by darinkrauss — APPROVED** [review]
Just one comment, but LGTM! ⚽
